File: 131_Palindrome_Partitioning.py
# Leetcode problem link : https://leetcode.com/problems/palindrome-partitioning/
# Time Complexity:    O(n*2^n)
# Space Complexity:   O(n) for Backtracking
#  Did this code successfully run on Leetcode : Yes
# Any problem you faced while coding this : No

# Your code here along with comments explaining your approach
'''
     1. Start from the first index and add to current path. Add current path to output at every instance
     2. If current string is palindrome then call backtracking recursively starting from next index and keep updating the path
     3. For loop will iterate through all elements in the input always starting from next index and going till end

'''

class Solution:
    def partition(self, s: str) -> List[List[str]]:
        output = []
        self.backtrack(s,output,0,[])
        return output
    
    def backtrack(self,s,output,index,path):
        if index >= len(s):
            temp = [x for x in path]
            output.append(temp)
            return
        
        for i in range(index,len(s)):
            if(self.isPalindrome(s[index:i+1])):
                path.append(s[index:i+1])
                self.backtrack(s,output,i+1,path)
                path.pop()
    
    def isPalindrome(self,s):
        if not s:
            return False
        
        l = 0
        r = len(s) - 1
        
        while l < r:
            if s[l] != s[r]:
                return False
            l += 1
            r -= 1
        return True
        

File: 131. Palindrome Partitioning.cpp
/*
Time: O(2^n)
Space: O(2^n)

Backtrack: For every call we have two choices, either choose (if it is a palindrome and proceed with a new empty substring) 
and not-choose (continue building the current substring)

*/

class Solution {
public:
    vector<vector<int>> is_pali;
    int n;
    string s;
    
    vector<vector<string>> v;
    

    //pre calculation of is_pali 2d array for all possible indices i,j
    void pre_calc(){
        is_pali = vector(n, vector(n, 0));

        for(int i=0;i<n;i++)  is_pali[i][i] = 1;

        for(int i=0;i<n-1;i++){
            for(int j=i+1;j<n;j++){
                is_pali[i][j] = 1;

                int lo = i, hi = j;
                while(lo<hi){
                    if(s[lo++]!=s[hi--])  is_pali[i][j] = 0;  
                }
            }
        }
    }

    vector<string> tmp;
    string str = "";
    void call(int curr, int last){
        if(curr == n){
            if(last == n)  v.push_back(tmp);
            return;
        }

        //not choose - continue
        str += s[curr];
        call(curr+1, last);
        str.pop_back(); //backtrack


        //choose if the substring is a palindrome - proceed with a new empty substring
        if(is_pali[last][curr]){
            str += s[curr];
            tmp.push_back(str);
            str = "";

            call(curr+1,curr+1);

            //backtrack
            str = tmp[tmp.size()-1]; 
            str.pop_back();
            tmp.pop_back();  

        }
    }

    vector<vector<string>> partition(string s) {
        n = s.length();
        this->s = s;
        pre_calc();
        call(0,0);

        return v;      
    }
};
